File: src/AiComposer.Core/State/StateStore.cs
using AiComposer.Core.Models;
using Microsoft.Data.Sqlite;

namespace AiComposer.Core.State;

/// <summary>
/// Persistent orchestrator state machine backed by SQLite.
/// Resumes from the last persisted state after a restart, preventing duplicate LLM calls.
/// </summary>
public sealed class StateStore : IDisposable
{
    private const int CircuitBreakerThreshold = 3;

    private readonly string _connectionString;

    /// <summary>Initialises the state store and creates the database schema if needed.</summary>
    public StateStore(string dbPath)
    {
        Directory.CreateDirectory(Path.GetDirectoryName(Path.GetFullPath(dbPath))!);
        _connectionString = $"Data Source={dbPath}";
        InitDb();
    }

    /// <summary>Returns the current state record for a ticket, or null if not tracked yet.</summary>
    public TicketStateRecord? GetTicket(string ticketId)
    {
        using var conn = OpenConnection();
        using var cmd = conn.CreateCommand();
        cmd.CommandText = "SELECT ticket_id, state, attempts, updated_at, last_error FROM ticket_state WHERE ticket_id = @id";
        cmd.Parameters.AddWithValue("@id", ticketId);
        using var reader = cmd.ExecuteReader();
        return reader.Read() ? ReadRecord(reader) : null;
    }

    /// <summary>Returns existing state or inserts a fresh record in <see cref="TicketState.WaitingForCoder"/>.</summary>
    public TicketStateRecord EnsureTicket(string ticketId)
    {
        var existing = GetTicket(ticketId);
        if (existing is not null) return existing;

        var now = DateTime.UtcNow.ToString("O");
        var record = new TicketStateRecord
        {
            TicketId = ticketId,
            State = TicketState.WaitingForCoder,
            Attempts = 0,
            UpdatedAt = now,
        };

        using var conn = OpenConnection();
        using var cmd = conn.CreateCommand();
        cmd.CommandText = "INSERT INTO ticket_state (ticket_id, state, attempts, updated_at, last_error) VALUES (@id, @state, @attempts, @updatedAt, @error)";
        cmd.Parameters.AddWithValue("@id", record.TicketId);
        cmd.Parameters.AddWithValue("@state", record.State);
        cmd.Parameters.AddWithValue("@attempts", record.Attempts);
        cmd.Parameters.AddWithValue("@updatedAt", record.UpdatedAt);
        cmd.Parameters.AddWithValue("@error", DBNull.Value);
        cmd.ExecuteNonQuery();

        return record;
    }

    /// <summary>
    /// Transitions a ticket to <paramref name="newState"/>.
    /// When <paramref name="incrementAttempt"/> is true, the attempt counter is incremented,
    /// and the circuit breaker kicks in after <see cref="CircuitBreakerThreshold"/> failures.
    /// </summary>
    public TicketStateRecord Transition(
        string ticketId,
        string newState,
        string? lastError = null,
        bool incrementAttempt = false)
    {
        var current = EnsureTicket(ticketId);
        var attempts = current.Attempts + (incrementAttempt ? 1 : 0);
        var resolvedState = attempts >= CircuitBreakerThreshold && incrementAttempt
            ? TicketState.PausedHumanIntervention
            : newState;

        var now = DateTime.UtcNow.ToString("O");
        using var conn = OpenConnection();
        using var cmd = conn.CreateCommand();
        cmd.CommandText =
            "UPDATE ticket_state SET state = @state, attempts = @attempts, updated_at = @updatedAt, last_error = @error WHERE ticket_id = @id";
        cmd.Parameters.AddWithValue("@state", resolvedState);
        cmd.Parameters.AddWithValue("@attempts", attempts);
        cmd.Parameters.AddWithValue("@updatedAt", now);
        cmd.Parameters.AddWithValue("@error", (object?)lastError ?? DBNull.Value);
        cmd.Parameters.AddWithValue("@id", ticketId);
        cmd.ExecuteNonQuery();

        return new TicketStateRecord
        {
            TicketId = ticketId,
            State = resolvedState,
            Attempts = attempts,
            UpdatedAt = now,
            LastError = lastError,
        };
    }

    /// <inheritdoc/>
    public void Dispose() => SqliteConnection.ClearAllPools();

    // ── private helpers ────────────────────────────────────────────────────────

    private SqliteConnection OpenConnection()
    {
        var conn = new SqliteConnection(_connectionString);
        conn.Open();
        return conn;
    }

    private void InitDb()
    {
        using var conn = OpenConnection();
        using var cmd = conn.CreateCommand();
        cmd.CommandText = """
            CREATE TABLE IF NOT EXISTS ticket_state (
                ticket_id  TEXT PRIMARY KEY,
                state      TEXT NOT NULL,
                attempts   INTEGER NOT NULL,
                updated_at TEXT NOT NULL,
                last_error TEXT
            )
            """;
        cmd.ExecuteNonQuery();
    }

    private static TicketStateRecord ReadRecord(SqliteDataReader reader) =>
        new()
        {
            TicketId = reader.GetString(0),
            State = reader.GetString(1),
            Attempts = reader.GetInt32(2),
            UpdatedAt = reader.GetString(3),
            LastError = reader.IsDBNull(4) ? null : reader.GetString(4),
        };
}

File: src/ai_composer/state_store.py
from __future__ import annotations

import sqlite3
from dataclasses import dataclass
from datetime import UTC, datetime
from pathlib import Path

from .models import TicketState


@dataclass(slots=True)
class TicketStateRecord:
    """Stored state for a ticket lifecycle."""

    ticket_id: str
    state: str
    attempts: int
    updated_at: str
    last_error: str | None = None


class StateStore:
    """Persistent orchestrator state machine backed by SQLite."""

    def __init__(self, db_path: str) -> None:
        self.db_path = db_path
        Path(db_path).parent.mkdir(parents=True, exist_ok=True)
        self._init_db()

    def _connect(self) -> sqlite3.Connection:
        return sqlite3.connect(self.db_path)

    def _init_db(self) -> None:
        with self._connect() as conn:
            conn.execute(
                """
                CREATE TABLE IF NOT EXISTS ticket_state (
                    ticket_id TEXT PRIMARY KEY,
                    state TEXT NOT NULL,
                    attempts INTEGER NOT NULL,
                    updated_at TEXT NOT NULL,
                    last_error TEXT
                )
                """
            )

    def get_ticket(self, ticket_id: str) -> TicketStateRecord | None:
        with self._connect() as conn:
            row = conn.execute(
                "SELECT ticket_id, state, attempts, updated_at, last_error FROM ticket_state WHERE ticket_id = ?",
                (ticket_id,),
            ).fetchone()
            if row is None:
                return None
            return TicketStateRecord(*row)

    def ensure_ticket(self, ticket_id: str) -> TicketStateRecord:
        existing = self.get_ticket(ticket_id)
        if existing:
            return existing
        now = datetime.now(tz=UTC).isoformat()
        record = TicketStateRecord(ticket_id, TicketState.WAITING_FOR_CODER, 0, now, None)
        with self._connect() as conn:
            conn.execute(
                """
                INSERT INTO ticket_state (ticket_id, state, attempts, updated_at, last_error)
                VALUES (?, ?, ?, ?, ?)
                """,
                (record.ticket_id, record.state, record.attempts, record.updated_at, record.last_error),
            )
        return record

    def transition(
        self,
        ticket_id: str,
        new_state: str,
        *,
        last_error: str | None = None,
        increment_attempt: bool = False,
    ) -> TicketStateRecord:
        current = self.ensure_ticket(ticket_id)
        attempts = current.attempts + (1 if increment_attempt else 0)
        if attempts >= 3 and increment_attempt:
            new_state = TicketState.PAUSED_HUMAN_INTERVENTION
        now = datetime.now(tz=UTC).isoformat()
        with self._connect() as conn:
            conn.execute(
                """
                UPDATE ticket_state
                SET state = ?, attempts = ?, updated_at = ?, last_error = ?
                WHERE ticket_id = ?
                """,
                (new_state, attempts, now, last_error, ticket_id),
            )
        return TicketStateRecord(ticket_id, new_state, attempts, now, last_error)

